alerting/notifier/channels/dingding.go

```diff
@@ -0,0 +1,159 @@
+package channels
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/url"
+
+	"github.com/prometheus/alertmanager/template"
+	"github.com/prometheus/alertmanager/types"
+)
+
+const defaultDingdingMsgType = "link"
+
+type dingDingSettings struct {
+	URL         string `json:"url,omitempty" yaml:"url,omitempty"`
+	MessageType string `json:"msgType,omitempty" yaml:"msgType,omitempty"`
+	Title       string `json:"title,omitempty" yaml:"title,omitempty"`
+	Message     string `json:"message,omitempty" yaml:"message,omitempty"`
+}
+
+func buildDingDingSettings(fc FactoryConfig) (*dingDingSettings, error) {
+	var settings dingDingSettings
+	err := json.Unmarshal(fc.Config.Settings, &settings)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal settings: %w", err)
+	}
+	if settings.URL == "" {
+		return nil, errors.New("could not find url property in settings")
+	}
+	if settings.MessageType == "" {
+		settings.MessageType = defaultDingdingMsgType
+	}
+	if settings.Title == "" {
+		settings.Title = DefaultMessageTitleEmbed
+	}
+	if settings.Message == "" {
+		settings.Message = DefaultMessageEmbed
+	}
+	return &settings, nil
+}
+
+func DingDingFactory(fc FactoryConfig) (NotificationChannel, error) {
+	n, err := newDingDingNotifier(fc)
+	if err != nil {
+		return nil, receiverInitError{
+			Reason: err.Error(),
+			Cfg:    *fc.Config,
+		}
+	}
+	return n, nil
+}
+
+// newDingDingNotifier is the constructor for the Dingding notifier
+func newDingDingNotifier(fc FactoryConfig) (*DingDingNotifier, error) {
+	settings, err := buildDingDingSettings(fc)
+	if err != nil {
+		return nil, err
+	}
+	return &DingDingNotifier{
+		Base:     NewBase(fc.Config),
+		log:      fc.Logger,
+		ns:       fc.NotificationService,
+		tmpl:     fc.Template,
+		settings: *settings,
+	}, nil
+}
+
+// DingDingNotifier is responsible for sending alert notifications to ding ding.
+type DingDingNotifier struct {
+	*Base
+	log      Logger
+	ns       WebhookSender
+	tmpl     *template.Template
+	settings dingDingSettings
+}
+
+// Notify sends the alert notification to dingding.
+func (dd *DingDingNotifier) Notify(ctx context.Context, as ...*types.Alert) (bool, error) {
+	dd.log.Info("sending dingding")
+
+	dingDingURL := buildDingDingURL(dd)
```

Comment:
Good to see that variable names for acronyms like "URL" are now following [Go conventions](https://github.com/golang/go/wiki/CodeReviewComments#initialisms)! :)